File: src/stresstesting/tamaf_java/src/Sender.java
package tamaf.stresstest;
import tamaf.*;
import java.util.Collections;

public class Sender extends Agent {
    public long triggerTime;
    public AgentID receiverAgentId;
    public Long rtt = null;
    public volatile boolean isRegistered = false;
    
    public Sender(String name, long triggerTime, String receiverIp, int receiverPort) {
        super(name);
        this.triggerTime = triggerTime;
        this.receiverAgentId = new AgentID("receiver", new Address(receiverIp, receiverPort));
    }
    
    private class SendBehavior extends SimpleBehavior {
        private boolean sent = false;
        private long sendTime = 0;
        
        public SendBehavior(Agent agent) {
            super(agent);
        }
        
        @Override
        public void action() {
            if (!sent) {
                long waitTime = ((Sender)agent).triggerTime - System.currentTimeMillis();
                if (waitTime > 0) {
                    ((Sender)agent).ams.scheduleWakeUp(waitTime);
                    return;
                }
                
                sendTime = System.currentTimeMillis();
                ACLMessage msg = new ACLMessage();
                msg.setSender(agent.getAgentDescription().getAgentid());
                msg.setPerformative(Performative.REQUEST);
                msg.addReceiver(((Sender)agent).receiverAgentId);
                msg.setContent(String.valueOf(sendTime));
                
                agent.send(msg);
                sent = true;
            }
            
            ACLMessageTemplate template = new ACLMessageTemplate();
            template.setPerformatives(Collections.singleton(Performative.INFORM));
            
            ACLMessage reply = receive(template); // throws EmptyReceiveException cooperatively
            
            long receiveTime = System.currentTimeMillis();
            ((Sender)agent).rtt = receiveTime - sendTime;
        }
        
        @Override
        public boolean done() {
            return ((Sender)agent).rtt != null;
        }
    }
    
    @Override
    public void setup(EMAInteraction emaInteraction) {
        this.isRegistered = true;
        addBehavior(new SendBehavior(this));
    }

    @Override
    public void stop() {
        super.stop();
        if (mts != null) {
            mts.stopMessageServer();
        }
    }
}

File: src/stresstesting/tamaf_python/src/sender.py
from tamaf import Agent, SimpleBehavior, ACLMessage, Performative, AgentID, Address, ACLMessageTemplate
import time

class SenderAgent(Agent):
    def __init__(self, name, trigger_time, receiver_ip, receiver_port):
        super().__init__(name)
        self.trigger_time = trigger_time
        self.receiver_agentid = AgentID(name="receiver", address=Address(receiver_ip, receiver_port))
        self.rtt = None
        
    class SendBehavior(SimpleBehavior):
        def __init__(self, agent):
            super().__init__()
            self.my_agent = agent
            self.sent = False

        def Action(self):
            if not self.sent:
                # Wait for trigger time by yielding
                if time.time() < self.my_agent.trigger_time:
                    return
                
                self.send_time = time.time()
                msg = ACLMessage(
                    sender=self.my_agent.agentDescription.agentid,
                    performative=Performative.REQUEST,
                    receiver=[self.my_agent.receiver_agentid],
                    content=str(self.send_time)
                )
                self.my_agent.Send(msg)
                self.sent = True
            
            # Now block wait for ACK (will raise EmptyReceiveException if none yet)
            reply = self.Receive(ACLMessageTemplate(performatives={Performative.INFORM}))
            receive_time = time.time()
            self.my_agent.rtt = receive_time - self.send_time

        def Done(self):
            return self.my_agent.rtt is not None
            
    def Setup(self, emainteraction):
        self.AddBehavior(self.SendBehavior(self))
